src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs

```diff
@@ -103,75 +105,139 @@ private static bool IsCodeActionNotSupportedByLSP(IUnifiedSuggestedAction sugges
         private static LSP.CodeAction[] GenerateCodeActions(
             CodeActionParams request,
             IUnifiedSuggestedAction suggestedAction,
-            LSP.CodeActionKind codeActionKind,
-            string currentTitle = "")
+            LSP.CodeActionKind codeActionKind)
         {
-            if (!string.IsNullOrEmpty(currentTitle))
+            var codeAction = suggestedAction.OriginalCodeAction;
+            var diagnosticsForFix = GetApplicableDiagnostics(request.Context, suggestedAction);
+
+            using var _ = ArrayBuilder<LSP.CodeAction>.GetInstance(out var builder);
+            using var _1 = ArrayBuilder<string>.GetInstance(out var codeActionPath);
+            var nestedCodeActions = CollectNestedActions(request, codeActionKind, diagnosticsForFix, suggestedAction, codeActionPath, isTopLevelCodeAction: true);
+
+            Command? nestedCodeActionCommand = null;
+            RoslynNestedCodeAction? nestedCodeAction = null;
+            var title = codeAction.Title;
+
+            var codeActionPathList = codeActionPath.ToArray();
+            if (nestedCodeActions.Any())
             {
-                // Adding a delimiter for nested code actions, e.g. 'Suppress or Configure issues|Suppress IDEXXXX|in Source'
-                currentTitle += '|';
+                nestedCodeAction = new RoslynNestedCodeAction()
+                {
+                    Title = title,
+                    Kind = codeActionKind,
+                    Diagnostics = diagnosticsForFix,
+                    NestedActions = nestedCodeActions,
+                    Data = new CodeActionResolveData(title, codeAction.CustomTags, request.Range, request.TextDocument, fixAllFlavors: null, nestedCodeAction: null, codeActionPathList)
+                };
+                nestedCodeActionCommand = new LSP.Command
+                {
+                    CommandIdentifier = CodeActionsHandler.RunNestedCodeActionCommandName,
+                    Title = title,
+                    Arguments = [new CodeActionResolveData(title, codeAction.CustomTags, request.Range, request.TextDocument, null, nestedCodeAction: nestedCodeAction, codeActionPathList)]
+                };
             }
 
-            var codeAction = suggestedAction.OriginalCodeAction;
-            currentTitle += codeAction.Title;
+            // We add an overarching action to the lightbulb that may contain nested actions.
+            // Selecting one of these actions from the list invokes a command on the client side to open
+            // a quick pick to select a nested action.
+            builder.Add(new LSP.CodeAction
+            {
+                Title = title,
+                Kind = codeActionKind,
+                Diagnostics = diagnosticsForFix,
+                Command = nestedCodeActionCommand,
+                Data = new CodeActionResolveData(title, codeAction.CustomTags, request.Range, request.TextDocument, fixAllFlavors: null, nestedCodeAction, codeActionPathList)
```

Comment:
So I was noticing we have 3 different `CodeActionResolveData` here which is a lot of duplication for one action (and a lot of duplicate serialization cost).
Is it possible to simplify this data structure?

For example - could we delete the `nestedCodeAction` entirely - and have the resolve data include the list `nestedCodeActions`?